pkg/cmd/pulumi/util_test.go

```diff
@@ -355,3 +356,119 @@ func TestStackLoadOption(t *testing.T) {
 		})
 	}
 }
+
+// TestGetUpdateMetadata tests that the update metadata is correctly populated
+// when running a Pulumi program.
+func TestPulumiCLIMetadata(t *testing.T) {
+	t.Parallel()
+
+	var stringFlag string
+	var intFlag int
+	var boolFlag bool
+
+	cmd := &cobra.Command{
+		Use: "my-cli-command",
+	}
+
+	cmd.PersistentFlags().StringVar(
+		&stringFlag, "name", "",
+		"This is a mock string flag that is set on the command.")
+
+	cmd.PersistentFlags().IntVar(
+		&intFlag, "age", 0,
+		"This is a mock int flag that is set on the command.")
+
+	cmd.PersistentFlags().BoolVar(
+		&boolFlag, "human", false,
+		"This is a mock boolean flag that is set on the command.")
+
+	actualEnv := map[string]string{}
+
+	getEnv := func() []string { // This is the function that returns the environment variables.
+		return []string{
+			// Checks that a normal flag is set in the environment.
+			"PULUMI_SKIP_UPDATE_CHECK=1",
+			// Checks that a flag is set in the environment to 'false'.
+			"PULUMI_EXPERIMENTAL=0",
+			// Checks that a non boolean flag set to true is 'set'.
+			"PULUMI_STRING_FLAG=true",
+			// Checks that an old and forgotten CLI flag is detected.
+			"PULUMI_DEPRECATED_FLAG=1.2.3",
+		}
+	}
+
+	subCmd := &cobra.Command{
```

Comment:
We should test that a declared flag that isn't set doesn't get added to the metadata. Also that being set to the default value _does_ still get added to the metadata.